python/cuml/cuml/accel/estimator_proxy.py

```diff
@@ -12,122 +12,328 @@
-        # The cuml signature may not align with the CPU signature.
-        # Additionally, some models support positional arguments.
-        # To work around this, we
-        # - Bind arguments to the CPU signature
-        # - Convert the arguments to named parameters
-        # - Translate them to cuml equivalents
-        # - Then forward them on to the cuml class
-        bound = self._cpu_model_sig.bind_partial(*args, **kwargs)
-        translated_kwargs, self._gpuaccel = self._hyperparam_translator(
-            **bound.arguments
-        )
-        super().__init__(**translated_kwargs)
-        self.build_cpu_model(**kwargs)
-
-    def __repr__(self):
-        return self._cpu_model.__repr__()
-
-    def __str__(self):
-        return self._cpu_model.__str__()
-
-    def __getattr__(self, attr):
-        # Don't dispatch __sklearn_clone__ so that cloning works as a
-        # as a regular estimator without __sklearn_clone__
-        may_dispatch = attr != "__sklearn_clone__"
-
-        if may_dispatch and hasattr(self._cpu_model_class, attr):
-            self.build_cpu_model()
-            self.gpu_to_cpu()
-            return getattr(self._cpu_model, attr)
-        raise AttributeError(
-            f"{type(self).__name__!r} object has no attribute {attr!r}"
-        )
+
+        # Forward some metadata to make the proxy class look like the CPU class
+        cls.__qualname__ = cls._cpu_class.__qualname__
+        cls.__doc__ = cls._cpu_class.__doc__
+        cls.__signature__ = inspect.signature(cls._cpu_class)
+        # We use the `_cpu_class_path` (which matches the patched path) instead of
+        # the original full module path so that the class (not an instance) can
+        # be pickled properly.
+        cls.__module__ = cls._gpu_class._cpu_class_path.rsplit(".", 1)[0]
+
+        # Add proxy method definitions for all public methods on CPU class
+        # that aren't already defined on the proxy class
+        methods = [
+            name
+            for name in dir(cls._cpu_class)
+            if not name.startswith("_")
+            and callable(getattr(cls._cpu_class, name))
+        ]
+
+        def _make_method(name):
+            cpu_method = getattr(cls._cpu_class, name)
+
+            @functools.wraps(cpu_method)
+            def method(self, *args, **kwargs):
+                return self._call_method(name, *args, **kwargs)
+
+            return method
+
+        for name in methods:
+            if not hasattr(cls, name):
+                setattr(cls, name, _make_method(name))
+
+    def _sync_params_to_gpu(self) -> None:
+        """Sync hyperparameters to GPU estimator from CPU estimator.
+
+        If the hyperparameters are unsupported, will cause proxy
+        to fallback to CPU until refit."""
+        if self._gpu is not None:
+            try:
+                params = self._gpu_class._params_from_cpu(self._cpu)
+                self._gpu.set_params(**params)
+            except UnsupportedOnGPU:
```

Comment:
Should we spit out some debug messages here based on the exception content? Might make it a bit easier for us to understand why we fallback to CPU during debug work. Perfectly fine to also deal with this in a follow-up when we look more closely at profiling.

Reply:
I thought about that (I think we could log any message contained in `UnsupportedOnGPU`, or surface that in the profiler somehow). I think that makes more sense in a follow-up though.